File: public/python/078-evaluate-reverse-polish-notation.py
# LeetCode Problem #150: Evaluate Reverse Polish Notation
# Difficulty: Medium
# Link: https://leetcode.com/problems/evaluate-reverse-polish-notation/

from typing import List

# ─────────────────────────────────────────────
# APPROACH 1: Stack-Based Evaluation | O(n) time | O(n) space
# EXPLAIN: Push numbers onto a stack; on encountering an operator, pop two operands and push result.
# WHEN: Classic stack application — standard interview answer.

def evalRPN_stack(tokens: List[str]) -> int:
    stack = []
    ops = {'+', '-', '*', '/'}

    for token in tokens:
        if token not in ops:
            stack.append(int(token))
        else:
            b, a = stack.pop(), stack.pop()
            if token == '+':
                stack.append(a + b)
            elif token == '-':
                stack.append(a - b)
            elif token == '*':
                stack.append(a * b)
            else:
                stack.append(int(a / b))  # truncate towards zero

    return stack[0]


# APPROACH 2: Stack with Lambda Map | O(n) time | O(n) space
# EXPLAIN: Use a dict of operator → lambda to simplify the conditional evaluation.
# WHEN: More concise; good when you want to demonstrate Python's first-class functions.

def evalRPN_lambda(tokens: List[str]) -> int:
    import operator
    ops = {
        '+': operator.add,
        '-': operator.sub,
        '*': operator.mul,
        '/': lambda a, b: int(a / b),
    }
    stack = []
    for token in tokens:
        if token in ops:
            b, a = stack.pop(), stack.pop()
            stack.append(ops[token](a, b))
        else:
            stack.append(int(token))
    return stack[0]


# APPROACH 3: Recursive Postfix Evaluation | O(n) time | O(n) space
# EXPLAIN: Walk tokens in reverse; recursively consume sub-expressions.
# WHEN: Demonstrates the recursive nature of expression trees; less practical but illuminating.

def evalRPN_recursive(tokens: List[str]) -> int:
    ops = {'+', '-', '*', '/'}
    idx = [len(tokens) - 1]  # mutable index

    def evaluate():
        token = tokens[idx[0]]
        idx[0] -= 1
        if token not in ops:
            return int(token)
        b = evaluate()
        a = evaluate()
        if token == '+': return a + b
        if token == '-': return a - b
        if token == '*': return a * b
        return int(a / b)

    return evaluate()


def evalRPN(tokens: List[str]) -> int:
    return evalRPN_stack(tokens)


# ── quick tests ──────────────────────────────────────────────────────────────
if __name__ == '__main__':
    cases = [
        (["2", "1", "+", "3", "*"], 9),
        (["4", "13", "5", "/", "+"], 6),
        (["10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"], 22),
    ]
    for fn in (evalRPN_stack, evalRPN_lambda, evalRPN_recursive, evalRPN):
        for tokens, expected in cases:
            assert fn(list(tokens)) == expected, f'{fn.__name__} mismatch'
    print('All tests passed.')

# Made with Bob

File: public/cpp/078-evaluate-reverse-polish-notation.cpp
/*
 * LeetCode Problem #150: Evaluate Reverse Polish Notation
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/evaluate-reverse-polish-notation/
 */

#include <vector>
#include <string>
#include <stack>
#include <unordered_set>
using namespace std;

class Solution {
public:
    // ==================== APPROACH 1: Stack-Based Evaluation ====================
    /*
     * Time Complexity: O(n)
     * Space Complexity: O(n)
     * EXPLAIN: Push numbers; on operator pop two operands, compute, push result.
     * WHEN: Classic stack application — standard interview answer.
     */
    int evalRPN_stack(vector<string>& tokens) {
        stack<long long> st;
        for (const string& token : tokens) {
            if (token == "+" || token == "-" || token == "*" || token == "/") {
                long long b = st.top(); st.pop();
                long long a = st.top(); st.pop();
                if      (token == "+") st.push(a + b);
                else if (token == "-") st.push(a - b);
                else if (token == "*") st.push(a * b);
                else                  st.push((long long)(a / b));
            } else {
                st.push(stoll(token));
            }
        }
        return (int)st.top();
    }

    // ==================== APPROACH 2: Stack with Lambda Map ====================
    /*
     * Time Complexity: O(n)
     * Space Complexity: O(n)
     * EXPLAIN: Use a map of operator → lambda to simplify the conditional evaluation.
     * WHEN: More concise; shows use of function objects.
     */
    int evalRPN_lambda(vector<string>& tokens) {
        unordered_map<string, function<long long(long long,long long)>> ops = {
            {"+", [](long long a, long long b){ return a + b; }},
            {"-", [](long long a, long long b){ return a - b; }},
            {"*", [](long long a, long long b){ return a * b; }},
            {"/", [](long long a, long long b){ return (long long)(a / b); }},
        };
        stack<long long> st;
        for (const string& token : tokens) {
            if (ops.count(token)) {
                long long b = st.top(); st.pop();
                long long a = st.top(); st.pop();
                st.push(ops[token](a, b));
            } else {
                st.push(stoll(token));
            }
        }
        return (int)st.top();
    }

    int evalRPN(vector<string>& tokens) {
        return evalRPN_stack(tokens);
    }
};
